specification/provider/providers.md

```diff
@@ -1,5 +1,159 @@
-# Providers
+# Provider
 
 ## Overview
 
-// TODO: this is a stub.
+The `provider` API defines interfaces that Provider Authors can use to abstract a particular flag management system, thus enabling the use of the `evaluation API` by Application Authors.
+
+#### Feature Provider Interface
+
+##### Requirement 2.1
+
+> The provider interface **MUST** define a `name` field or accessor, which identifies the provider implementation.
+
+#### Flag Value Resolution
+
+`Providers` are implementations of the `feature provider` interface, which may wrap vendor SDKs, REST API clients, or otherwise resolve flag values from the runtime environment.
+
+##### Requirement 2.2
+
+> The `feature provider` interface **MUST** define methods to resolve flag values, with parameters `flag key` (string, required), `default value` (boolean | number | string | structure, required), `evaluation context` (optional), and `evaluation options` (optional), which returns a `flag resolution` structure.
+
+```
+// example flag resolution function
+resolveBooleanValue(flagKey, defaultValue, context, options);
+```
+
+See: [flag resolution structure](../types.md#flag-resolution), [flag value resolution](../glossary.md#flag-value-resolution)
+
+##### Condition 2.3
+
+> The implementing language type system differentiates between strings, numbers, booleans and structures.
+>
+> ##### Conditional Requirement 2.3.1
+>
+> > The `feature provider` interface **MUST** define methods for typed flag resolution, including boolean, numeric, string, and structure.
+
+```
+// example boolean flag value resolution
+FlagResolution resolveBooleanValue(string flagKey, boolean defaultValue, context: EvaluationContext, options: FlagEvaluationOptions);
+
+// example string flag value resolution
+FlagResolution resolveStringValue(string flagKey, string defaultValue, context: EvaluationContext, options: FlagEvaluationOptions);
+
+// example number flag value resolution
+FlagResolution resolveNumberValue(string flagKey, number defaultValue, context: EvaluationContext, options: FlagEvaluationOptions);
+
+// example structure flag value resolution
+FlagResolution resolveStructureValue(string flagKey, JsonObject defaultValue, context: EvaluationContext, options: FlagEvaluationOptions);
+```
+
+##### Requirement 2.4
+
+> In cases of normal execution, the `provider` **MUST** populate the `flag resolution` structure's `value` field with the resolved flag value.
+
+##### Requirement 2.4
+
+> In cases of abnormal execution, the `provider` **MUST** populate the `flag resolution` structure's `value` field with the default flag value.
```

Comment:
In Requirement 2.8 we say that a provider is allowed to throw during abnormal execution. Saying they MUST also populate the default value seems to contradict that somewhat.

Reply:
You're right, this is really the job of the client (caller) in this case.